File: studio/src/studio/contracts.ts
// Settle a contract book against the modeled spot price.
//
// The TypeScript twin of src/power_dispatch/contracts.py, and the two carry the
// same arithmetic on purpose: a supplier who marks a scenario in the browser and
// then re-runs it in a notebook has to read the same pesos. contracts.test.ts
// checks the browser against worked numbers, and tests/test_contracts.py checks
// Python against the same ones.

import type { GridKey } from '../lib/types'
import type { ChronoHour } from './chrono'

export const GRIDS: GridKey[] = ['luzon', 'visayas', 'mindanao']

export interface Contract {
  name?: string
  grid: GridKey
  mw: number
  strike_php_kwh: number
  /** buy fixes what the holder pays, sell fixes what the holder receives */
  side?: 'buy' | 'sell'
  /** hours the contract covers, 0 to 23. Absent means the whole day */
  hours?: number[]
}

export interface SettledContract {
  name: string
  grid: GridKey
  mw: number
  strike: number
  side: 'buy' | 'sell'
  hoursCovered: number
  mwh: number
  spotValue: number
  strikeValue: number
  position: number
  meanSpot: number
}

export interface OpenRow {
  grid: GridKey
  loadMw: number
  openMwh: number
  openCost: number
  coveredPct: number
}

export interface Settlement {
  contracts: SettledContract[]
  position: number
  open: OpenRow[]
  openCost: number
}

/** Every problem with a book, as messages a person can act on. */
export function validateBook(book: Contract[]): string[] {
  const e: string[] = []
  book.forEach((c, i) => {
    if (!GRIDS.includes(c.grid)) e.push(`Contract ${i + 1} needs a grid.`)
    if (!Number.isFinite(c.mw) || c.mw < 0)
      e.push(`Contract ${i + 1} needs volume in MW, and it cannot be negative.`)
    if (!Number.isFinite(c.strike_php_kwh))
      e.push(`Contract ${i + 1} needs a strike price in PhP/kWh.`)
    if (c.hours && c.hours.some((h) => h < 0 || h > 23))
      e.push(`Contract ${i + 1} names an hour outside 0 to 23.`)
  })
  return e
}

/**
 * Mark a book against one solved day.
 *
 * One MW held for one hour is one MWh, and a price in PhP/kWh is a thousand
 * pesos per MWh, so one contract-hour is worth mw * 1000 * price.
 */
export function settle(
  hours: ChronoHour[],
  book: Contract[],
  loadMw: Partial<Record<GridKey, number>> = {}
): Settlement {
  const n = hours.length
  const rows: SettledContract[] = book.map((c) => {
    const cover = (c.hours ?? Array.from({ length: n }, (_, i) => i)).filter((h) => h < n)
    const side = c.side ?? 'buy'
    const sign = side === 'buy' ? 1 : -1
    const spot = cover.reduce((s, h) => s + hours[h].price[c.grid] * c.mw * 1000, 0)
    const strike = c.strike_php_kwh * c.mw * 1000 * cover.length
    return {
      name: c.name || `${c.grid} ${c.mw} MW at P${c.strike_php_kwh}`,
      grid: c.grid,
      mw: c.mw,
      strike: c.strike_php_kwh,
      side,
      hoursCovered: cover.length,
      mwh: c.mw * cover.length,
      spotValue: spot,
      strikeValue: strike,
      position: sign * (spot - strike),
      meanSpot: cover.length
        ? cover.reduce((s, h) => s + hours[h].price[c.grid], 0) / cover.length
        : 0,
    }
  })

  // only a buy contract covers a load. A sell contract is a separate position.
  const covered: Record<string, number[]> = {}
  for (const g of GRIDS) covered[g] = new Array(n).fill(0)
  book.forEach((c) => {
    if ((c.side ?? 'buy') !== 'buy') return
    for (const h of c.hours ?? Array.from({ length: n }, (_, i) => i))
      if (h < n) covered[c.grid][h] += c.mw
  })

  const open: OpenRow[] = []
  for (const g of GRIDS) {
    const mw = loadMw[g]
    if (!mw) continue
    let mwh = 0
    let cost = 0
    for (let h = 0; h < n; h++) {
      const gap = Math.max(0, mw - covered[g][h])
      mwh += gap
      cost += gap * 1000 * hours[h].price[g]
    }
    open.push({
      grid: g,
      loadMw: mw,
      openMwh: mwh,
      openCost: cost,
      coveredPct: n ? 100 * (1 - mwh / (mw * n)) : 0,
    })
  }

  return {
    contracts: rows,
    position: rows.reduce((s, r) => s + r.position, 0),
    open,
    openCost: open.reduce((s, r) => s + r.openCost, 0),
  }
}

export interface PositionChange {
  base: Settlement
  scenario: Settlement
  positionChange: number
  openCostChange: number
  netChange: number
}

/** What one scenario does to the position, which is the question people ask. */
export function comparePosition(
  baseHours: ChronoHour[],
  scenarioHours: ChronoHour[],
  book: Contract[],
  loadMw: Partial<Record<GridKey, number>> = {}
): PositionChange {
  const base = settle(baseHours, book, loadMw)
  const scenario = settle(scenarioHours, book, loadMw)
  const positionChange = scenario.position - base.position
  const openCostChange = scenario.openCost - base.openCost
  return {
    base,
    scenario,
    positionChange,
    openCostChange,
    netChange: positionChange - openCostChange,
  }
}

/** A starting book, so the view opens on something a reader can change. */
export const SAMPLE_BOOK: Contract[] = [
  {
    name: 'Power supply agreement',
    grid: 'luzon',
    mw: 250,
    strike_php_kwh: 6.4,
    side: 'buy',
  },
  {
    name: 'Evening peak block',
    grid: 'luzon',
    mw: 100,
    strike_php_kwh: 9.0,
    side: 'buy',
    hours: [18, 19, 20, 21],
  },
]

File: src/power_dispatch/contracts.py
"""Settle a contract book against the modeled spot price.

A retail supplier, a plant owner, and an industrial buyer all ask the same
question about any scenario: what does it do to my position? The dispatch model
answers half of it, because it produces an hourly spot price per grid. The other
half is arithmetic on a contract book, and the book is the caller's.

One contract is a volume at a strike price over a set of hours:

    {"name": "PSA with the DU", "grid": "luzon", "mw": 250,
     "strike_php_kwh": 6.4, "hours": [18, 19, 20, 21], "side": "buy"}

`side` says which way the position runs. A **buy** contract fixes the price the
holder pays, so a spot price above the strike is a gain against buying at spot.
A **sell** contract fixes the price the holder receives, so the same spot move is
a loss. Omit `hours` and the contract covers all 24.

`settle` returns pesos, per contract and in total, and never guesses at volume
the book does not carry. It also reports the open position: the load a caller
declares and has not contracted, priced at spot.

What this is not: no credit terms, no take-or-pay, no capacity fee, no
line-rental or wheeling charge, and no tax. It marks energy against modeled spot
and nothing else. A settlement statement has more lines than this.

    import power_dispatch as pd
    r = pd.run_scenario({"date": "2026-06-17", "opts": {"fuel_avail_delta":
                        {"luzon": {"coal": -647}}}})
    pos = pd.settle(r, book, load_mw={"luzon": 400})
"""

from __future__ import annotations

GRIDS = ("luzon", "visayas", "mindanao")
SIDES = ("buy", "sell")


def validate_book(book: list) -> list[str]:
    """Every problem with a contract book, as messages a person can act on."""
    e: list[str] = []
    if not isinstance(book, list):
        return ["the contract book must be a list"]
    for i, c in enumerate(book):
        if not isinstance(c, dict):
            e.append(f"contract[{i}] must be an object")
            continue
        if c.get("grid") not in GRIDS:
            e.append(f"contract[{i}].grid must be one of {', '.join(GRIDS)}")
        for f in ("mw", "strike_php_kwh"):
            v = c.get(f)
            if not isinstance(v, (int, float)) or isinstance(v, bool):
                e.append(f"contract[{i}].{f} must be a number")
        if c.get("mw", 0) < 0:
            e.append(f"contract[{i}].mw must not be negative. Use side to pick direction")
        side = c.get("side", "buy")
        if side not in SIDES:
            e.append(f"contract[{i}].side must be buy or sell")
        hrs = c.get("hours")
        if hrs is not None:
            if not isinstance(hrs, list) or not hrs:
                e.append(f"contract[{i}].hours must be a non-empty list of hours")
            elif any(not isinstance(h, int) or h < 0 or h > 23 for h in hrs):
                e.append(f"contract[{i}].hours must hold whole hours from 0 to 23")
        if "name" in c and not isinstance(c["name"], str):
            e.append(f"contract[{i}].name must be a string")
    return e


def settle(result: dict, book: list, load_mw: dict | None = None) -> dict:
    """Mark a contract book against one solved day.

    `result` is what run_scenario returns. `load_mw` is the caller's own load
    per grid in MW, flat across the day, and it decides the open position.

    Every figure is pesos for the day. One MW held for one hour is one MWh, and
    a price in PhP/kWh is a thousand pesos per MWh, so the peso value of one
    contract-hour is mw * 1000 * price.
    """
    problems = validate_book(book)
    if problems:
        raise ValueError("contract book has problems:\n  " + "\n  ".join(problems))
    hours = result["hours"]
    n = len(hours)

    rows = []
    for c in book:
        g = c["grid"]
        cover = c.get("hours") or list(range(n))
        cover = [h for h in cover if h < n]
        mw = float(c["mw"])
        strike = float(c["strike_php_kwh"])
        sign = 1.0 if c.get("side", "buy") == "buy" else -1.0
        spot_cost = sum(hours[h]["price"][g] * mw * 1000.0 for h in cover)
        strike_cost = strike * mw * 1000.0 * len(cover)
        # a buy contract is worth what spot would have cost minus what it fixed
        rows.append(
            {
                "name": c.get("name") or f"{g} {mw:g} MW at P{strike:g}",
                "grid": g,
                "mw": mw,
                "strike_php_kwh": strike,
                "side": c.get("side", "buy"),
                "hours_covered": len(cover),
                "mwh": mw * len(cover),
                "spot_value_php": round(spot_cost, 2),
                "strike_value_php": round(strike_cost, 2),
                "position_php": round(sign * (spot_cost - strike_cost), 2),
                "mean_spot_php_kwh": round(
                    sum(hours[h]["price"][g] for h in cover) / len(cover), 4
                )
                if cover
                else 0.0,
            }
        )

    covered = {g: [0.0] * n for g in GRIDS}
    for c, r in zip(book, rows):
        if c.get("side", "buy") != "buy":
            continue
        for h in c.get("hours") or range(n):
            if h < n:
                covered[c["grid"]][h] += float(c["mw"])

    open_rows = []
    for g, mw in (load_mw or {}).items():
        if g not in GRIDS:
            continue
        open_mwh = 0.0
        open_cost = 0.0
        for h in range(n):
            gap = max(0.0, float(mw) - covered[g][h])
            open_mwh += gap
            open_cost += gap * 1000.0 * hours[h]["price"][g]
        open_rows.append(
            {
                "grid": g,
                "load_mw": float(mw),
                "open_mwh": round(open_mwh, 2),
                "open_cost_php": round(open_cost, 2),
                "covered_share_pct": round(
                    100.0 * (1 - open_mwh / (float(mw) * n)) if mw and n else 0.0, 1
                ),
            }
        )

    return {
        "date": result.get("summary", {}).get("date"),
        "hours": n,
        "contracts": rows,
        "position_php": round(sum(r["position_php"] for r in rows), 2),
        "open": open_rows,
        "open_cost_php": round(sum(r["open_cost_php"] for r in open_rows), 2),
        "note": (
            "Energy marked against modeled spot. No capacity fee, no wheeling "
            "charge, no tax, and no credit terms."
        ),
    }


def compare(base: dict, scenario: dict, book: list, load_mw: dict | None = None) -> dict:
    """What one scenario does to the position, which is the question people ask.

    Settles the same book twice and reports the change. A trip, a price move, or
    a new data center shows up here as pesos rather than as a price chart.
    """
    a = settle(base, book, load_mw)
    b = settle(scenario, book, load_mw)
    by_name = {r["name"]: r for r in a["contracts"]}
    moves = []
    for r in b["contracts"]:
        was = by_name.get(r["name"])
        if not was:
            continue
        moves.append(
            {
                "name": r["name"],
                "grid": r["grid"],
                "base_position_php": was["position_php"],
                "scenario_position_php": r["position_php"],
                "change_php": round(r["position_php"] - was["position_php"], 2),
                "base_mean_spot": was["mean_spot_php_kwh"],
                "scenario_mean_spot": r["mean_spot_php_kwh"],
            }
        )
    return {
        "base": a,
        "scenario": b,
        "contracts": moves,
        "position_change_php": round(b["position_php"] - a["position_php"], 2),
        "open_cost_change_php": round(b["open_cost_php"] - a["open_cost_php"], 2),
        "net_change_php": round(
            (b["position_php"] - a["position_php"])
            - (b["open_cost_php"] - a["open_cost_php"]),
            2,
        ),
        "note": a["note"],
    }
